File: compiler/internal/vm/toolchain/test/autotest/aw/taskpool.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Copyright (c) 2024 Huawei Device Co., Ltd.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Description: A task pool that can execute tasks asynchronously.
"""

import asyncio
from queue import Queue
from threading import Thread, current_thread
from time import time


class TaskPool(object):
    def __init__(self):
        self.task_queue = Queue()
        self.event_loop = None
        self.task_exception = None
        self.event_loop_thread = None
        self._start_event_loop()

    def submit(self, coroutine, callback=None):
        # add item to the task queue
        self._task_add()
        future = asyncio.run_coroutine_threadsafe(coro=coroutine, loop=self.event_loop)
        future.add_done_callback(callback) if callback else None
        # remove item from the task queue after the task is done
        future.add_done_callback(self._task_done)

    def await_taskpool(self):
        asyncio.run_coroutine_threadsafe(coro=self._stop_loop(), loop=self.event_loop)

    def task_join(self):
        self.task_queue.join()

    def _task_add(self, item=1):
        self.task_queue.put(item)

    def _task_done(self, future):
        # clear the task queue and stop the task pool once an exception occurs in the task
        if future.exception():
            while not self.task_queue.empty():
                self.task_queue.get()
                self.task_queue.task_done()
            self.task_exception = future.exception()
            return
        self.task_queue.get()
        self.task_queue.task_done()

    def _set_and_run_loop(self, loop):
        self.event_loop = loop
        asyncio.set_event_loop(loop)
        loop.run_forever()

    async def _stop_loop(self, interval=1):
        # wait for all tasks in the event loop is done, then we can close the loop
        while True:
            if self.task_queue.empty():
                self.event_loop.stop()
                return
            await asyncio.sleep(interval)

    def _start_event_loop(self):
        loop = asyncio.new_event_loop()
        self.event_loop_thread = Thread(target=self._set_and_run_loop, args=(loop,))
        self.event_loop_thread.daemon = True
        # Specifies the thread name to be able to save log of thread to the module_run.log file
        self.event_loop_thread.name = current_thread().name + "-" + str(time()).replace('.', '')[-5:]
        self.event_loop_thread.start()

File: compiler/internal/vm/common_runtime/common_components/taskpool/taskpool.cpp
/*
 * Copyright (c) NeXTHub Corporation. All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * Author: Tunjay Akbarli
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Please contact NeXTHub Corporation, 651 N Broad St, Suite 201,
 * Middletown, DE 19709, New Castle County, USA.
 */

#include "common_components/taskpool/taskpool.h"

#include "common_components/platform/cpu.h"

namespace common {
Taskpool *Taskpool::GetCurrentTaskpool()
{
    static Taskpool *taskpool = new Taskpool();
    return taskpool;
}

void Taskpool::Initialize(int threadNum, std::function<void(os::thread::NativeHandleType)> prologueHook,
                          const std::function<void(os::thread::NativeHandleType)> epilogueHook)
{
    std::lock_guard<std::mutex> guard(mutex_);
    if (isInitialized_++ <= 0) {
        runner_ = std::make_unique<Runner>(TheMostSuitableThreadNum(threadNum), prologueHook, epilogueHook);
    }
}

void Taskpool::Destroy(int32_t id)
{
    std::lock_guard<std::mutex> guard(mutex_);
    if (isInitialized_ <= 0) {
        return;
    }
    isInitialized_--;
    if (isInitialized_ == 0) {
        runner_->TerminateThread();
    } else {
        runner_->TerminateTask(id, TaskType::ALL);
    }
}

void Taskpool::TerminateTask(int32_t id, TaskType type)
{
    if (isInitialized_ <= 0) {
        return;
    }
    runner_->TerminateTask(id, type);
}

uint32_t Taskpool::TheMostSuitableThreadNum(uint32_t threadNum) const
{
    if (threadNum > 0) {
        return std::min<uint32_t>(threadNum, MAX_TASKPOOL_THREAD_NUM);
    }
    uint32_t numOfThreads = std::min<uint32_t>(NumberOfCpuCore() / 2, MAX_TASKPOOL_THREAD_NUM);
    if (numOfThreads > MIN_TASKPOOL_THREAD_NUM) {
        return numOfThreads - 1;  // 1 for daemon thread.
    }
    return MIN_TASKPOOL_THREAD_NUM;  // At least MIN_TASKPOOL_THREAD_NUM GC threads, and 1 extra daemon thread.
}

void Taskpool::ForEachTask(const std::function<void(Task *)> &f)
{
    if (isInitialized_ <= 0) {
        return;
    }
    runner_->ForEachTask(f);
}
}  // namespace common

